python/venv/private/venv.bzl

```diff
@@ -214,11 +210,65 @@ def _py_venv_binary_impl(ctx):
             env_inherit = [],
             targets = ctx.attr.data,
         ),
-        OutputGroupInfo(
-            python_zip_file = depset([python_zip_file]),
+        PyMainInfo(
+            main = ctx.file.main,
+            srcs = ctx.files.srcs,
         ),
     ]
 
+def _py_venv_zipapp_impl(ctx):
+    venv_toolchain = ctx.toolchains[venv_common.TOOLCHAIN_TYPE]
+    py_info = ctx.attr.binary[PyInfo]
+    main_info = ctx.attr.binary[PyMainInfo]
+
+    python_zip_file = venv_common._create_python_zip_file(
+        ctx = ctx,
+        venv_toolchain = venv_toolchain,
+        py_info = py_info,
+        main = _compute_main(
+            ctx = ctx,
+            main = main_info.main,
+            srcs = main_info.srcs,
+        ),
+        runfiles = ctx.attr.binary[DefaultInfo].default_runfiles,
+        files_to_run = ctx.attr.binary[DefaultInfo].files_to_run,
+    )
+
+    return DefaultInfo(
+        files = depset([python_zip_file]),
+    )
+
+py_venv_zipapp = rule(
+    doc = """\
+A `py_venv_zipapp` is an executable Python zipapp which contains all of the
+dependencies and runfiles for a given executable.
+
+```python
+load("@rules_venv//python/venv:defs.bzl", "py_venv_binary", "py_venv_zipapp")
+
+py_venv_binary(
+    name = "foo",
+    srcs = ["foo.py"],
+)
+
+py_venv_zipapp(
+    name = "foo_pyz",
+    binary = ":foo",
+)
+```
+""",
+    implementation = _py_venv_zipapp_impl,
+    attrs = {
```

Comment:
Add `shebang` as an optional attribute?

Reply:
Added, you can see it is applied:
```
sputt@sputt-mbp2 rules_venv % unzip -l bazel-bin/python/venv/private/tests/python_zip_file/with_runfiles/writer_pyz.pyz > /dev/null
warning [bazel-bin/python/venv/private/tests/python_zip_file/with_runfiles/writer_pyz.pyz]:  10 extra bytes at beginning or within zipfile
  (attempting to process anyway)
```
```
sputt@sputt-mbp2 rules_venv % head -1 bazel-bin/python/venv/private/tests/python_zip_file/with_runfiles/writer_pyz.pyz
#!python3
```

Also fixed the zip writer to add the new line.